api-report/driver-definitions.api.md

```diff
@@ -134,7 +134,7 @@ export interface IDocumentDeltaConnectionEvents extends IErrorEvent {
     // (undocumented)
     (event: "op", listener: (documentId: string, messages: ISequencedDocumentMessage[]) => void): any;
     // (undocumented)
-    (event: "signal", listener: (message: ISignalMessage) => void): any;
+    (event: "signal", listener: (message: ISignalMessage | ISignalMessage[]) => void): any;
```

Comment:
i think this change is probably fine, but i want to ensure is aligns with where the protocol is going. 

For ops, it seems we are moving away from the client sending arrays. I think this is the correct path, as it blocks any enumeration/fan-out on the server at the protocol, 1 message === 1 envelope. For ops, do we think the server would want to send array to the client? Seems possible, but i don't think there are any immediate plans here.

For signals, i think we should take the same approach that ops take, in the client never sends an array, again to block any enumeration/fan-out on the server at the protocol. I also think it's fine for the server to batch this, and send an array to the client. @GaryWilber do you see a case where the server would batch message, even after clients batch signals within those messages? Similar to ops it seems possible, the server is also in a better position to correctly manage payload sizes.

Curious what @vladsud  thinks about this.

Reply:
Do you have any docs about clients moving away from sending arrays?

The current system:
- Clients can send an array of op batches in a single "submitOp" event to the server and the server can send back the sequenced array of ops in a single "op" event. The server can also choose to split it into multiple "op" events if needed (to ensure we stay within some payload limitations across kafka / websockets).
- Clients can send an array of signals in a single "submitSignal' event to the server and the server **must** send back the signals in individual "signal" events - This means a single submitSignal event that includes multiple signals always fans out into multiple signal events 
     - This PR changes this so the server has the ability to send multiple signals back to the client in a single signal event.

Reply:
>>> Clients can send an array of signals in a single "submitSignal' event to the server and the server must send back the signals in individual "signal" events - This means a single submitSignal event that includes multiple signals always fans out into multiple signal events
This PR changes this so the server has the ability to send multiple signals back to the client in a single signal event.

is this true? i don't see anywhere in our code where clients send arrays of signals.

Reply:
The [submitSignal handler](https://github.com/microsoft/FluidFramework/blob/ad43659bb657918148119b81782ea13d02834af2/server/routerlicious/packages/lambdas/src/alfred/index.ts#L769) on the server side supports arrays. Maybe clients aren't leveraging that?

Reply:
ok. so we do send an array, but its not leveraged anywhere above the call to send: 
https://github.com/microsoft/FluidFramework/blame/a627d4034d83c6b4d4908a1e70f73bb1c6e650cc/packages/drivers/driver-base/src/documentDeltaConnection.ts#L279

Reply:
I have slightly different question: Is this shape for compat only and we plan to remove single signal from signature in the future? If so, when can it happen? can it happen soon for both FRS & ODSP stacks, or are we getting stuck here for years as that's how long it will take for this change to happen for FRS?

Reply:
I think it would be nice to remove single signal from the signature to make it so the signature matches op's (only support arrays).
I assume we (FRS/ODSP) would only be able to remove the old signature once all host apps update their runtimes to a version that supports the array model. I assume we'll be stuck in this state for a while, somewhat similar to the single-commit work.

Reply:
i don't think there is a shortcut here. we support both for a while. once all clients have the ability to process array, we enable on server, and drop the single message path.

Reply:
in theory we could still push the signature forward, and for the details into the driver. i think that would work here.

Reply:
Re-opening this PR since I do want to get this in. 
@anthony-murphy Do you think the current iteration is good or should I try to push the signals array signature forward? I know you approved, but I want to double check since it's been a few months.